src/cargo/ops/cargo_package/vcs/gix.rs

```diff
@@ -0,0 +1,408 @@
+use crate::core::{Package, Workspace};
+use crate::ops::cargo_package::vcs::{GitVcsInfo, VcsInfo};
+use crate::ops::PackageOpts;
+use crate::sources::PathEntry;
+use crate::{CargoResult, GlobalContext};
+use anyhow::Context;
+use cargo_util::paths;
+use gix::bstr::ByteSlice;
+use gix::dir::walk::EmissionMode;
+use gix::index::entry::Mode;
+use gix::status::tree_index::TrackRenames;
+use gix::worktree::stack::state::ignore::Source;
+use std::collections::HashSet;
+use std::path::{Path, PathBuf};
+use tracing::debug;
+
+/// Checks if the package source is in a *git* DVCS repository.
+///
+/// If *git*, and the source is *dirty* (e.g., has uncommitted changes),
+/// and `--allow-dirty` has not been passed,
+/// then `bail!` with an informative message.
+/// Otherwise, return the sha1 hash of the current *HEAD* commit,
+/// or `None` if no repo is found.
+#[tracing::instrument(skip_all)]
+pub fn check_repo_state(
+    p: &Package,
+    src_files: &[PathEntry],
+    ws: &Workspace<'_>,
+    opts: &PackageOpts<'_>,
+) -> CargoResult<Option<VcsInfo>> {
+    let gctx = ws.gctx();
+    let Ok(repo) = gix::discover(p.root()) else {
+        gctx.shell().verbose(|shell| {
+            shell.warn(format_args!(
+                "no (git) VCS found for `{}`",
+                p.root().display()
+            ))
+        })?;
+        // No Git repo found. Have to assume it is clean.
+        return Ok(None);
+    };
+
+    let Some(workdir) = repo.workdir() else {
+        debug!(
+            "no (git) workdir found for repo at `{}`",
+            repo.path().display()
+        );
+        // No git workdir. Have to assume it is clean.
+        return Ok(None);
+    };
+
+    debug!("found a git repo at `{}`", workdir.display());
+    let path = p.manifest_path();
+
+    let manifest_exists = path.exists();
+    let path = paths::strip_prefix_canonical(path, workdir).unwrap_or_else(|_| path.to_path_buf());
+    let rela_path =
+        gix::path::to_unix_separators_on_windows(gix::path::os_str_into_bstr(path.as_os_str())?);
+    if !manifest_exists {
+        gctx.shell().verbose(|shell| {
+            shell.warn(format_args!(
+                "Cargo.toml not found at `{}` in workdir `{}`",
+                path.display(),
+                workdir.display()
+            ))
+        })?;
+        // TODO: Either remove this whole block, or have a test.
+        //       It's hard to have no Cargo.toml here?
```

Comment:
It's unclear to me how this could possibly happen - there is no test for it either. From looking at `git blame` it wasn't obvious what motivated this check.
Probably nothing wrong with keeping it.

Reply:
Maybe only some parts of the package is under git? Like `src` is a git repo.